File: round1B 2008/mousetrap/fenwick.py
'''
Created on Jun 10, 2013

@author: Yubin Bai
'''


class FenwickTree:

    '''
    // In this implementation, the tree is represented by a list
    // Elements are numbered by 0, 1, ..., n-1.
    // tree[i] is sum of elements with indexes i&(i+1)..i, inclusive.
    // (Note: this is a bit different from what is proposed in Fenwick's article.
    // To see why it makes sense, think about the trailing 1's in binary
    // representation of indexes.)
    '''

    def __init__(self, size, initValue=0):
        self.tree = [initValue] * size

    def increase(self, i, delta):
        '''
         Increases value of i-th element by ''delta''.
        '''
        while i < len(self.tree):
            self.tree[i] += delta
            i |= i + 1

    def getsum(self, left, right):
        '''
        left and right must be inside the range
        Returns sum of elements with indexes left..right, inclusive; (zero-based);
        '''
        if right >= left:
            return self._sum(right) - self._sum(left - 1)
        else:
            return self._sum(len(self.tree) - 1) - self._sum(left - 1) + \
                self._sum(right)

    def _sum(self, index):
        currSum = 0
        while index >= 0:
            currSum += self.tree[index]
            index &= index + 1
            index -= 1
        return currSum

File: fenwick.jl
type FenwickTree{T}

    tree::Array{T, 1}

    function FenwickTree{T}(size::Int, v::T)
        a = 1
        while a <= size
            a = a << 1
        end
        new([v for i in 1:a])
    end

end

FenwickTree(size, v) = FenwickTree{typeof(v)}(size::Int, v)

function increase{T}(t::FenwickTree{T}, idx::Int, delta::T)
    while idx + 1 < length(t.tree)
        t.tree[idx + 1] += delta
        idx |= idx + 1
    end
end

function getSum{T}(t::FenwickTree{T}, left::Int, right::Int)
    if right >= left
        _sum(t, right) - _sum(t, left - 1)
    else
        throw(BoundsError())
    end
end

function _sum{T}(t::FenwickTree{T}, idx)
    currSum = 0
    while idx >= 0
        currSum += t.tree[idx + 1]
        idx &= idx + 1
        idx -= 1
    end
    return currSum
end

N = 10000000

t = FenwickTree(N, 0.0)
println("typeof(t) $(typeof(t))")


for i in 1:N
    increase(t, i, i*1.0)
end

@time for i in 1:N
    BigFloat(getSum(t, 1, i))
end

# println(getSum(t, 100, 1))

